// mand-mobile/lib/tab-picker/index.js
;(function(){
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', '../popup', '../popup/title-bar', '../tabs', '../icon', '../radio', '../_util', './style/index.css'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('../popup'), require('../popup/title-bar'), require('../tabs'), require('../icon'), require('../radio'), require('../_util'), require('./style/index.css'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.popup, global.titleBar, global.tabs, global.icon, global.radio, global._util, global.index);
    global.index = mod.exports;
  }
})(this, function (exports, _popup, _titleBar, _tabs, _icon, _radio, _util) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _popup2 = _interopRequireDefault(_popup);

  var _titleBar2 = _interopRequireDefault(_titleBar);

  var _tabs2 = _interopRequireDefault(_tabs);

  var _icon2 = _interopRequireDefault(_icon);

  var _radio2 = _interopRequireDefault(_radio);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  var _components;

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  exports.default = {
    name: 'md-tab-picker',

    components: (_components = {}, _defineProperty(_components, _popup2.default.name, _popup2.default), _defineProperty(_components, _titleBar2.default.name, _titleBar2.default), _defineProperty(_components, _tabs2.default.name, _tabs2.default), _defineProperty(_components, _icon2.default.name, _icon2.default), _defineProperty(_components, _radio2.default.name, _radio2.default), _components),

    props: {
      value: {
        type: Boolean,
        default: false
      },
      title: {
        type: String,
        default: ''
      },
      okText: {
        type: String,
        default: '确认'
      },
      cancelText: {
        type: String,
        default: '取消'
      },
      data: {
        type: Array,
        default: function _default() {
          return [];
        }
      },
      dataStruct: {
        type: String,
        default: 'normal'
      },
      defaultIndex: {
        type: Array,
        default: function _default() {
          return [];
        }
      },
      optionRender: {
        type: Array,
        default: function _default() {
          return [];
        },
        validator: function validator(value) {
          if (value.length > 0) {
            return value.every(function (item) {
              return typeof item === 'function';
            });
          } else {
            return true;
          }
        }
      },
      asyncFunc: {
        type: Function,
        default: function _default() {
          return _util.noop;
        }
      }
    },

    data: function data() {
      return {
        isTabPickerShow: false,
        subTitles: [],
        renderData: [],
        defaultTabIndex: 0,
        currentIndex: 0,
        isLoading: true,
        isDataError: false,
        currentColumnLock: false
      };
    },


    watch: {
      value: function value(val) {
        val && (this.isTabPickerShow = val);
      },
      isTabPickerShow: function isTabPickerShow(val) {
        !val && this.$emit('input', val);
      },

      data: {
        handler: function handler(val, oldVal) {
          if (!(0, _util.compareObjects)(val, oldVal)) {
            this.$_initTabPicker();
          }
        },

        deep: true
      },
      isDataError: function isDataError(val) {
        var _this = this;

        if (val) {
          setTimeout(function () {
            _this.isDataError = false;
          }, 2000);
        }
      }
    },

    computed: {
      hasTitleSlotScope: function hasTitleSlotScope() {
        return !!this.$scopedSlots.titles;
      },
      hasOptionSlotScope: function hasOptionSlotScope() {
        return !!this.$scopedSlots.option;
      }
    },

    created: function created() {
      this.$_initTabPicker();
    },


    methods: {
      $_initTabPicker: function $_initTabPicker() {
        switch (this.dataStruct) {
          case 'normal':
            this.$_initNoCascadeData();
            break;
          case 'cascade':
            this.$_initCascadeData();
            break;
          case 'async':
            this.$_initAsyncCascadeData();
            break;
          default:
            break;
        }
      },
      $_initCascadeData: function $_initCascadeData() {
        var _this2 = this;

        if (this.data.length === 0) {
          return;
        }
        this.isLoading = false;
        if (this.defaultIndex && this.defaultIndex.length > 0) {
          var i = 0;
          var func = function func(array) {
            if (i < _this2.defaultIndex.length) {
              var temp = _this2.defaultIndex[i];
              array.forEach(function (item, eq) {
                if (eq === temp) {
                  _this2.subTitles.push(item.label);
                  _this2.renderData.push({ index: i, clickedKey: temp, data: array });
                  if (item && item.children && Array.isArray(item.children)) {
                    i++;
                    func(item.children);
                  }
                }
              });
            } else {
              _this2.defaultTabIndex = i;
              return false;
            }
          };
          func(this.data);
        } else {
          this.subTitles.push('请选择');
          this.renderData.push({ index: 0, clickedKey: -1, data: this.data });
        }
      },
      $_initNoCascadeData: function $_initNoCascadeData() {
        var _this3 = this;

        if (this.data.length === 0) {
          return;
        }
        this.isLoading = false;
        this.data.forEach(function (item, index) {
          var temp = {
            index: index,
            clickedKey: _this3.defaultIndex.length > 0 && ~_this3.defaultIndex[index] ? _this3.defaultIndex[index] : -1,
            data: item.children
          };
          _this3.renderData.push(temp);
          var currentColumn = _this3.renderData[index];
          if (_this3.defaultIndex && _this3.defaultIndex.length > 0) {
            _this3.subTitles.push(currentColumn.data[currentColumn.clickedKey].label);
          } else {
            _this3.subTitles.push(item.label);
          }
        });
      },
      $_initAsyncCascadeData: function $_initAsyncCascadeData() {
        this.asyncFunc(null, this.$_renderNextTabContent);
      },
      $_renderNextTabContent: function $_renderNextTabContent(err, data) {
        var _this4 = this;

        this.isLoading = false;
        if (err) {
          this.isDataError = err;
          return;
        }
        try {
          this.subTitles.splice(this.currentIndex + 1, this.subTitles.length - 1, '请选择');
          this.renderData.splice(this.currentIndex + 1, this.renderData.length - 1, {
            index: this.currentIndex,
            clickedKey: -1,
            data: data.options,
            asyncFunc: data.asyncFunc
          });
          if (this.renderData.length > 1) {
            this.$nextTick(function () {
              _this4.$refs.tabs.selectTab(++_this4.currentIndex);
            });
          }
        } catch (error) {
          this.isDataError = true;
        }
      },
      $_onShow: function $_onShow() {
        this.$emit('show');
      },
      $_onHide: function $_onHide() {
        this.$emit('hide');
      },
      $_onConfirm: function $_onConfirm() {
        this.isTabPickerShow = false;
        var selectedItem = this.getSelectedItem();
        this.$emit('confirm', selectedItem);
      },
      $_onCancel: function $_onCancel() {
        this.isTabPickerShow = false;
        this.$emit('cancel');
      },
      $_onRadioChange: function $_onRadioChange(value, index) {
        var _this5 = this;

        if (this.dataStruct === 'cascade' && this.currentColumnLock) {
          return;
        }
        this.currentColumnLock = true;
        this.subTitles[this.currentIndex] = value.label;
        var currentColumn = this.renderData[this.currentIndex];
        currentColumn.clickedKey = index;
        this.$emit('change', {
          selectTab: this.currentIndex,
          selectIndex: index,
          selectItem: value
        });
        if (this.dataStruct === 'cascade') {
          if (value && value.children && Array.isArray(value.children) && value.children.length > 0) {
            this.subTitles.splice(this.currentIndex + 1, this.subTitles.length - 1, '请选择');
            this.renderData.splice(this.currentIndex + 1, this.renderData.length - 1, {
              index: ++this.currentIndex,
              clickedKey: -1,
              data: value.children
            });
            this.$nextTick(function () {
              _this5.$refs.tabs.selectTab(_this5.currentIndex);
              setTimeout(function () {
                _this5.currentColumnLock = false;
              }, 500);
            });
            return;
          }
          this.currentColumnLock = false;
        }
        if (this.dataStruct === 'async' && currentColumn.asyncFunc) {
          value = _extends({
            index: index
          }, value);

          this.isLoading = true;
          typeof currentColumn.asyncFunc === 'function' && currentColumn.asyncFunc(value, this.$_renderNextTabContent);
        }
      },
      $_onIndexChange: function $_onIndexChange(index) {
        this.currentIndex = index;
      },
      getSelectedItem: function getSelectedItem() {
        return this.renderData.map(function (item, index) {
          if (~item.clickedKey) {
            var selected = item.data[item.clickedKey];
            return {
              index: index,
              item: {
                label: selected.label,
                value: selected.value
              }
            };
          } else {
            return null;
          }
        });
      }
    }
  };
});
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"md-tab-picker"},[_c('md-popup',{attrs:{"position":"bottom","mask-closable":false},on:{"show":_vm.$_onShow,"hide":_vm.$_onHide},model:{value:(_vm.isTabPickerShow),callback:function ($$v) {_vm.isTabPickerShow=$$v},expression:"isTabPickerShow"}},[_c('md-popup-title-bar',{attrs:{"title":_vm.title,"ok-text":_vm.okText,"cancel-text":_vm.cancelText},on:{"confirm":_vm.$_onConfirm,"cancel":_vm.$_onCancel}}),_vm._v(" "),_c('div',{staticClass:"md-tab-picker-content"},[_c('md-tabs',{ref:"tabs",attrs:{"titles":_vm.subTitles,"default-index":_vm.defaultTabIndex,"force-use-array":_vm.hasTitleSlotScope},on:{"indexChanged":_vm.$_onIndexChange},scopedSlots:_vm._u([{key:"title",fn:function(props){return [_vm._t("titles",null,{label:props})]}}])},_vm._l((_vm.subTitles),function(title,index){return _c('div',{key:index},[_c('md-radio',{key:_vm.renderData[index].clickedKey,ref:"radio1",refInFor:true,attrs:{"options":_vm.renderData[index].data,"default-index":~_vm.renderData[index].clickedKey ? _vm.renderData[index].clickedKey : -1,"is-slot-scope":_vm.hasOptionSlotScope},on:{"change":_vm.$_onRadioChange},scopedSlots:_vm._u([{key:"default",fn:function(props){return [_vm._t("option",null,{option:props.option,index:index})]}}])})],1)})),_vm._v(" "),(_vm.isLoading || _vm.isDataError)?_c('div',{staticClass:"slot-wrapper"},[_c('div',{staticClass:"slot-inner"},[(_vm.isDataError)?_vm._t("error",[_vm._v("数据异常")]):_vm._e(),_vm._v(" "),(_vm.isLoading)?_vm._t("loading",[_vm._v("loading")]):_vm._e()],2)]):_vm._e()],1)],1)],1)}
__vue__options__.staticRenderFns = []
__vue__options__._scopeId = "data-v-5d4e08a1"
if (module.hot) {(function () {  var hotAPI = require("vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-5d4e08a1", __vue__options__)
  } else {
    hotAPI.reload("data-v-5d4e08a1", __vue__options__)
  }
})()}
